### ISSUE TITLE
Updating from torch 2.0 to 2.1/2.2/2.3/2.4/2.5 results in gradient explosion with SDPA and bf16 

### ISSUE THREAD

**Issue by ssharpe42** [pytorch/pytorch#141241]
### 🐛 Describe the bug

We are training a huggingface transformer using pytorch native SDPA. We recently upgraded from torch 2.0, and found that **this change alone** caused our training runs to diverge. 

We cannot upload any images and unfortunately are unable to share code since it is internal/proprietary. The training loss is almost identical across torch versions for over half the run and then they diverge. The diverges look similar to this issue https://github.com/pytorch/pytorch/issues/139298, however grad_norm can sometimes get up to values near 100.

Here are some comparisons before and after different changes (all with grad clipping and max grad norm of 1.0):

Torch 2.0 + bf16 + SDPA + lr=5e-5 + gradient accum = 2--> clean loss curve :white_check_mark:
Torch **2.1/2.2/2.3/2.4/2.5** + bf16 + SDPA + lr=5e-5 +  gradient accum = 2 --> exploding gradients :x:
Torch 2.4 + bf16 + SDPA + lr=5e-5 + **gradient accum = 1** -->  exploding gradients :x:
Torch 2.4 +  **fp16** + SDPA + lr=5e-5 + gradient accum = 2 --> clean loss curve :white_check_mark:
Torch 2.5 +  fp16 + SDPA + **lr=1e-4** + gradient accum = 2 -->exploding gradients :x:
Torch 2.4 + bf16 + No SPDA (eager) + lr=5e-5 + gradient accum = 2 --> clean loss curve :white_check_mark: 

We also tried what[ this issue](
https://github.com/pytorch/pytorch/issues/139298) reported as a solution (`torch.backends.cuda.enable_cudnn_sdp(False)`) but this also resulted in exploding gradients.

We are unsure whether this is an issue with SDPA after 2.0 or just finicky pre-training. Is there anything that changed post torch 2.0 that could be related to this issue? Seems odd that upgrading torch alone would cause this shift in behavior. Any help would be appreciated!

### Versions

This is our environment after upgrading to torch 2.4

```
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.15 | packaged by conda-forge | (main, Oct 16 2024, 01:24:24) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.10.226-214.880.amzn2.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 550.118
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               96
On-line CPU(s) list:                  0-95
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            2
Stepping:                             7
BogoMIPS:                             6000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            1.5 MiB (48 instances)
L1i cache:                            1.5 MiB (48 instances)
L2 cache:                             48 MiB (48 instances)
L3 cache:                             71.5 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-23,48-71
NUMA node1 CPU(s):                    24-47,72-95
Vulnerability Gather data sampling:   Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:          KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Vulnerable
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] torch==2.4.0+cu121
[pip3] torchaudio==2.4.0+cu121
[pip3] torchvision==0.19.0+cu121
[pip3] triton==3.0.0
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.5.1.17                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] torch                     2.4.0+cu121              pypi_0    pypi
[conda] torchaudio                2.4.0+cu121              pypi_0    pypi
[conda] torchvision               0.19.0+cu121             pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
```

**Comment by ssharpe42** [comment]
@drisspg @malfet I have narrowed this down to an issue with a custom bias we are adding to the attention scores. In both flex_attention and sdpa we see gradients explode but in eager we do not.  

We are using an alibi bias score_mod where with some custom bias implementations.  Specifically, my data is a sequence of events where each event is associated with an event time, when I use the implementation for alibi that leverages the absolute time difference instead of the position in my flex attention and sdpa implementation the gradients tend to explode but with eager they don't.  

| Attention Implementation   | Alibi implementation | Gradient explosion | 
| -------- | ------- | ------- |
| eager  | position   | no |
| eager | event time  | no |
| sdpa  | position   | <running> |
| sdpa  | event time     | yes |
| flex_attention  | position   | no |
| flex_attention  | event time     | yes |

Other details
- Training the model using DDP and fp16 
- I have tried clipping the values of the time differences to make sure these numbers aren't large but it still ends up with gradient explosion 

Are there any other related issues I should be looking at for this problem? 

See example below for how it is used. In the eager and sdpa versions we just add the values to the attention mask. 

```
from torch.nn.attention.flex_attention import (
    _score_mod_signature,
    flex_attention,
)

flex_attention = torch.compile(flex_attention, dynamic=False)

def generate_alibi_bias(
    num_heads: int, device: torch.device, event_time: torch.FloatTensor
) -> _score_mod_signature:

    # Normal position based alibi (this works in flex_attention)
    def alibi_mod(score, b, h, q_idx, kv_idx):
        slope = torch.exp2(-((h + 1) * 8.0 / num_heads))
        bias = -abs(kv_idx - q_idx) * slope
        return score + bias

    # This fails and results in exploding gradients in flex_attention but not eager
    def alibi_time_mod(score, b, h, q_idx, kv_idx):
        slope = torch.exp2(-((h + 1) * 8.0 / num_heads))
        bias = -torch.abs(event_time[b][kv_idx] - event_time[b][q_idx]) * slope
        return score + bias

    return alibi_time_mod


N_BATCH = 1
N_HEAD = 12
SEQ_LEN = 4
D_QK = 64

torch.set_default_device("cuda")

def main():

    event_time = torch.tensor([[1.0, 10.0, 20.0, 100.0]])
    q = torch.randn(N_BATCH, N_HEAD, SEQ_LEN, D_QK, requires_grad=True)
    k = torch.randn(N_BATCH, N_HEAD, SEQ_LEN, D_QK, requires_grad=True)
    v = torch.randn(N_BATCH, N_HEAD, SEQ_LEN, D_QK, requires_grad=True)
    score_mod = generate_alibi_bias(N_HEAD, device=q.device, event_time=event_time)
    flex_attention(q, k, v, score_mod=score_mod).sum().backward()
```

**Comment by drisspg** [comment]
Hey Im going to close this as this is quite issue and I am unable to reproduce on main / the code has changed quite a bit since then
